src/self-managed-debug/src/main.rs

```diff
@@ -0,0 +1,722 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Debug tool for self managed environments.
+
+use std::fmt::Debug;
+
+use std::process;
+
+use std::sync::LazyLock;
+
+use chrono::{Duration, Utc};
+use clap::Parser;
+use k8s_openapi::api::apps::v1::{Deployment, ReplicaSet, StatefulSet};
+use k8s_openapi::api::networking::v1::{NetworkPolicy, NetworkPolicyPeer, NetworkPolicyPort};
+use k8s_openapi::apimachinery::pkg::util::intstr::IntOrString;
+use kube::config::KubeConfigOptions;
+use kube::{Client, Config};
+use tabled::{Style, Table, Tabled};
+
+use mz_build_info::{build_info, BuildInfo};
+
+use mz_ore::cli::{self, CliConfig};
+
+use k8s_openapi::api::core::v1::{ContainerStatus, Pod, Service};
+
+use kube::api::{Api, ListParams, LogParams, ObjectMeta};
+use mz_ore::error::ErrorExt;
+
+use std::fs::File;
+use std::io::Write;
+
+pub const BUILD_INFO: BuildInfo = build_info!();
+pub static VERSION: LazyLock<String> = LazyLock::new(|| BUILD_INFO.human_version(None));
+
+#[derive(Parser, Debug)]
+#[clap(name = "self-managed-debug", next_line_help = true, version = VERSION.as_str())]
+pub struct Args {
+    // === Kubernetes options. ===
+    #[clap(long, env = "KUBERNETES_CONTEXT")]
+    kubernetes_context: Option<String>,
+    #[clap(long, value_delimiter = ',', num_args = 1..)]
+    kubernetes_namespaces: Vec<String>,
+}
+
+#[tokio::main]
+async fn main() {
+    let args: Args = cli::parse_args(CliConfig {
+        env_prefix: Some("SELF_MANAGED_DEBUG_"),
+        enable_version_flag: true,
+    });
+
+    if let Err(err) = run(args).await {
+        eprintln!(
+            "self-managed-debug: fatal: {}\nbacktrace: {}",
+            err.display_with_causes(),
+            err.backtrace()
+        );
+        process::exit(1);
+    }
+}
+
+async fn run(args: Args) -> Result<(), anyhow::Error> {
+    let client = create_k8s_client(args.kubernetes_context.clone()).await?;
+    // TODO: Make namespaces mandatory
+    // TODO: Print a warning if namespace doesn't exist
+    for namespace in args.kubernetes_namespaces {
```

Comment:
We probably want to store things in an organized directory structure. Maybe something like:
`materialize-debug-{iso_datetime}/{namespace}/{resource_type}/{resource_name}.yaml`

or, in the case of logs:
`materialize-debug-{iso_datetime}/{namespace}/logs/{pod_name}.{current_or_previous}.log`

Reply:
That makes sense, even if we're zipping at the end!